src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/kafka_client_cache.h"
+
+#include "pandaproxy/client_cache_error.h"
+#include "pandaproxy/logger.h"
+
+#include <chrono>
+
+using namespace std::chrono_literals;
+
+namespace pandaproxy {
+
+static constexpr auto clean_timer_period = 10s;
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg,
+  size_t max_size,
+  std::vector<config::broker_authn_endpoint> kafka_api,
+  model::timestamp::type keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _kafka_has_sasl{false}
+  , _keep_alive{keep_alive} {
+    // Is there a Kafka listener with SASL enabled?
+    auto ep_it = std::find_if(
+      kafka_api.begin(),
+      kafka_api.end(),
+      [](const config::broker_authn_endpoint& ep) {
+          return ep.authn_method == config::broker_authn_method::sasl;
+      });
+
+    _kafka_has_sasl = ep_it != kafka_api.end();
+
+    _clean_timer.set_callback([this] {
+        // Do not need a gate since there are limited items
+        // in the cache. Therefore, relative to other design,
+        // garbage collection is inexpensive here.
+        clean_stale_clients();
+    });
+    _clean_timer.arm_periodic(clean_timer_period);
+}
+
+kafka_client_cache::~kafka_client_cache() { _clean_timer.cancel(); }
+
+client_ptr kafka_client_cache::fetch(credential_t user) {
+    auto& inner_hash = _cache.get<underlying_hash>();
+    auto it_hash = inner_hash.find(user.name);
+
+    if (it_hash == inner_hash.end()) {
+        return nullptr;
+    }
+
+    // Otherwise user-client found.
+
+    // Convert the hash iterator to list iterator
+    auto it_list = _cache.project<underlying_list>(it_hash);
+    auto& inner_list = _cache.get<underlying_list>();
+
+    // Extract the item. This is OK because elements are
+    // not copied or moved. Instead internal pointers
+    // are simply re-balanced and no allocations are made.
+    auto node = inner_list.extract(it_list);
+
+    // Update the timestamp and put it to the front
+    node.value().client->last_used = model::timestamp::now();
+    inner_list.push_front(node.value());
```

Comment:
Should still use a gate/mutex here just-in-case multiple fetches for the same user occur concurrently.

This scenario will be covered in future scale tests where multiples users will issue rest requests.